[FSDP] New rate limiter and memory reuse system

kwen2501 opened the pull request:
### 1. Motivation
Today's FSDP uses CPU-GPU synchronization to limit issuance rate of all-gather. This stalls the CPU thread from issuance new tasks when waiting for GPU computation to finish. See "CPU Thread Activities".
<img width="910" alt="Screenshot 2023-09-28 at 1 08 16 AM" src="https://github.com/pytorch/pytorch/assets/6676466/1d07138f-b8fc-4e06-8bfb-69b8575c2df4">

And when the CPU thread resumes, there is an 235 us overhead between the time the synchronization is met and the time a new all-gather is launched. See "Overhead (zoomed in)". This overhead would turn into GPU utilization gap.

<img width="1011" alt="Screenshot 2023-09-28 at 1 12 24 AM" src="https://github.com/pytorch/pytorch/assets/6676466/d381ea12-c5b3-4c2b-aa96-deb33475f409">

### 2. Proposal
Replace CPU synchronization with stream wait in limitation of all-gather issuance. 

We ask the unshard stream to wait for the resharding event of the "-2" all-gather, which marks the compute completion of the "-2" FSDP instance.

In order for the CUDA caching allocator to work with the stream wait, we ask it to hold the all-gather buffer till the "+2" all-gather. We do so by stashing the all-gather buffer into the existing "free event queue", and free it only when we are about to issue the "+2" all-gather (after "+2" all-gather has established wait relationship).

CPU timeline as follows:
![fsdp_rate_limiter](https://github.com/pytorch/pytorch/assets/6676466/a5089bc0-d9ab-4345-b12e-7ae3c4a62b43)

### 2.1 Additional Changes
**2.1.1 Moved `_free_storage` from reshard to +2 unshard.**
![free_storage](https://github.com/pytorch/pytorch/assets/6676466/7b912f65-95bf-4494-bbaf-fd7f2945e5b4)

**2.1.2 Consolidated `BACKWARD_PRE` and `BACKWARD_POST` options for backward prefetch.**
i.e. we would just use BACKWARD_POST, for example, the completion of layer 12 backward would trigger all-gather of layer 10 (with a stream wait, of course)

### 2.2 Backward wait system
The backward wait is implemented as follows:
![backward](https://github.com/pytorch/pytorch/assets/6676466/47e9e5fc-ff2b-4494-968f-4f0d75086dbc)

Location 1:
We need to allocate two buffers to achieve compute-communication overlap. This is auto-satisfied by the state of the free-event queue at the end of forward -- which still has two last layers stashed. In fact, by such stashing, we save two all-gathers (for the last two layers) compared to today's FSDP. 

Location 2:
What happen here are:
Compute on unshard buffer B done -> Unshard stream wait for compute stream -> Free unshard buffer B -> Switch to unshard stream -> Allocate unshard buffer B+1 -> Perform all-gather on B+1

Location 3:
Switch back to default stream, call `split_and_view` to get individual parameter -> Default stream wait for unshard stream. We switch back to default stream so that the backward op for `split_and_view` would be also on default stream. Otherwise, autograd would use a separate stream for this backward op and call `record_stream` on the param tensors.

### 3. Evaluation
![380210060_682184966891666_4956717331232594368_n](https://github.com/pytorch/pytorch/assets/6676466/7b67e75e-068b-49d7-bf18-e0ff7206b6ea)

Cc: @awgu @albanD @janeyx99

kwen2501 committed:
New rate limiter

kwen2501 committed:
Lint

kwen2501 committed:
Add clear() method to FreeEventQueue

kwen2501 committed:
Move limit AG logic up

kwen2501 committed:
Free storage when clearing queue

kwen2501 committed:
Comments

kwen2501 committed:
Clear is unused now

albanD commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -1634,8 +1667,26 @@ def reshard(self, free_unsharded_flat_param: bool):
+                # unshard stream). We keep the tensor alive via stashing.
+                comp_done_event = self._device_handle.Event()
+                comp_done_event.record()
+                unsharded_flat_param = self._get_padded_unsharded_flat_param()
```
Note that any Tensor that is a view into the storage will prevent the de-allocation (even if it looks at only a subset of the storage). 
So here you can take whichever Tensor that has a view into that storage so we should use whichever is cheapest to get.
The "padded" in the name here might mean that this is an expensive one to get?

kwen2501 replied:
Thanks for sharing the knowledge!

`_get_padded_unsharded_flat_param` is called to get a reference to the actual unshard buffer used -- FSDP creates two such buffers: a "regular" one and a "full-precision" one. I guess the additional one was created to support mixed precision. At runtime, only one of these two unshard buffers is used based on some knobs. I guess the mixed precision support could have been done in a different way, but figured better not to touch it in this PR.

albanD reviewed (commented):
Sounds good to me. What's the plan for testing / make sure it behaves as we want?

kwen2501 committed:
Rename to _prefetch_next_handle

kwen2501 committed:
Clear queue after forward; backward now works

kwen2501 committed:
Consolidate BACKWARD_PRE and BACKWARD_POST for prefetch

kwen2501 committed:
Clear queue after forward if in eval mode

kwen2501 committed:
Allocate buffer on default stream for backward

kwen2501 committed:
Change FreeEventQueue to OrderedDict

kwen2501 committed:
Add in_forward and in_backward property

kwen2501 committed:
Remove EventWithTensor

kwen2501 committed:
Direct free; fix root handle setting

kwen2501 committed:
Move view() op to default stream

kwen2501 committed:
Allocate backward unshard buffer in unshard stream

kwen2501 committed:
Unify fwd and bwd unshard

kwen2501 committed:
Lint

kwen2501 committed:
Comments

kwen2501 committed:
Fix rebase: device_mesh error

awgu commented:
`distributed/fsdp/test_fsdp_unshard_params.py::TestUnshardParams::test_unshard_params_respects_reshard`
```
FSDP(
  nn.Sequential(
      FSDP(nn.Linear),
      nn.Linear,
)
```
There are two `FlatParameter`s: the root's, referred to as `outer_flat_param`, and the first linear's, referred to as `inner_flat_param`. The assertion failing is saying that the inner flat parameter is not freed after forward.
https://github.com/pytorch/pytorch/blob/2a40b7efcb273d1689b55763fade49263adcc788/test/distributed/fsdp/test_fsdp_unshard_params.py#L314
The existing behavior was that only the root module's flat parameter (in this case, `outer_flat_param`) should not be freed after forward for `ShardingStrategy.FULL_SHARD`.

---
`distributed/fsdp/test_fsdp_comm.py::TestCommunication::test_communication_nested_model_False_use_no_sync_False_sharding_strategy_None`
This test checks that FSDP issues the expected number of all-gathers. It looks like you now issue 2 fewer all-gathers (expected 19 but got 17).
https://github.com/pytorch/pytorch/blob/2a40b7efcb273d1689b55763fade49263adcc788/test/distributed/fsdp/test_fsdp_comm.py#L282
This is probably related to the 1st test.

---
`distributed/fsdp/test_wrap.py::TestFSDPWrap::test_main_wrap_api_cpu_offload_CPUOffload(offload_params=False)_backward_prefetch_BackwardPrefetch_BACKWARD_PRE_forward_prefetch_False_cuda_init_mode_CUDAInitMode_CUDA_AFTER`
This test is probably erroring because you clamp the `backward_prefetch` to `BACKWARD_POST` whereas the test expects that the passed-in `backward_prefetch` is respected.
https://github.com/pytorch/pytorch/blob/2a40b7efcb273d1689b55763fade49263adcc788/test/distributed/fsdp/test_wrap.py#L369

kwen2501 committed:
Adapt test: update num all_gathers

kwen2501 committed:
Add more FSDP layers to overcome 'respect_reshard' test

kwen2501 committed:
Fix test: backward prefetch option

kwen2501 committed:
Update test_wrap to use BACKWARD_POST

kwen2501 committed:
[c10d] Pass avoidRecordStreams into collective() function

kwen2501 committed:
Revert changes in _base functions

kwen2501 committed:
Add stream-stream wait for backward compute and reduce-scatter

kwen2501 committed:
Lint

kwen2501 committed:
Fix NoneType.wait issue

kwen2501 committed:
Fix checkpoint CI error

Add `pause` and `resume` methods to prevent recomputation of Activation
Checkpointing from getting into forward pass's free event queue
